### PR TITLE
Improve error logging for ReportingService

### PR DESCRIPTION
**Pull request by xshipeng**
The SEVERE level log message in ReportingService will only appear after 5 minutes of continuous connection failure.

### CODE REVIEW

File: src/main/java/com/wavefront/sdk/common/logging/DelegatingLogger.java [hunk 1]
```diff
@@ -43,7 +43,7 @@ public void log(LogRecord logRecord) {
    * This is a JDK8-specific implementation that is quite expensive because it fetches the
    * current stack trace. TODO: switch to StackWalker after migrating to JDK9+
    */
-  private void inferCaller(LogRecord logRecord) {
+   void inferCaller(LogRecord logRecord) {
```
Comment (hanwavefront) [inline src/main/java/com/wavefront/sdk/common/logging/DelegatingLogger.java:46]: nit: formatting

File: src/main/java/com/wavefront/sdk/common/clients/service/ReportingService.java [hunk 2]
```diff
@@ -128,16 +130,18 @@ private int safeGetResponseCodeAndClose(HttpURLConnection urlConn) {
     try {
       statusCode = urlConn.getResponseCode();
     } catch (IOException ex) {
-      log.log(Level.SEVERE, "Unable to obtain status code from the Wavefront service at "
-          + urlConn.getURL().toString(), ex);
+      MESSAGE_SUPPRESSING_LOGGER.log(urlConn.getURL().toString(), Level.SEVERE,
+          "Unable to obtain status code from the Wavefront service at "
+              + urlConn.getURL().toString());
```
Comment (hanwavefront) [inline src/main/java/com/wavefront/sdk/common/clients/service/ReportingService.java:135]: Let's log the exception message to provide some additional context.  E.g., ` + " due to: " + ex.getMessage()`

File: src/main/java/com/wavefront/sdk/common/clients/service/ReportingService.java [hunk 3]
```diff
@@ -128,16 +130,18 @@ private int safeGetResponseCodeAndClose(HttpURLConnection urlConn) {
...
     }
 
     try {
       readAndClose(urlConn.getErrorStream());
     } catch (IOException ex) {
-      log.log(Level.SEVERE, "Unable to read and close error stream from the Wavefront service at "
-          + urlConn.getURL().toString(), ex);
+      MESSAGE_SUPPRESSING_LOGGER.log(urlConn.getURL().toString(), Level.SEVERE,
+          "Unable to read and close error stream from the Wavefront service at "
+              + urlConn.getURL().toString());
```
Comment (hanwavefront) [inline src/main/java/com/wavefront/sdk/common/clients/service/ReportingService.java:144]: Ditto, let's log the exception message.

File: src/main/java/com/wavefront/sdk/common/logging/MessageSuppressingLogger.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    for (final StackTraceElement stackTraceElement : stackTraceElements) {
+      String className = stackTraceElement.getClassName();
+      String methodName = stackTraceElement.getMethodName();
+      if (lookingForLogMethod) {
+        // Locate the log method and then find the caller.
+        if (className.equals(currentClassName) && methodName.equals(logMethodName)) {
+          lookingForLogMethod = false;
+        }
+      } else {
+        logRecord.setSourceClassName(className);
```
Comment (hanwavefront) [inline src/main/java/com/wavefront/sdk/common/logging/MessageSuppressingLogger.java:73]: Does the infer work correctly if you call `logger.severe(msg)`?  For this case, I think we might still need the if statement that's here in the parent method.  Let's add an assert to the tests to check the source class and method name too.

**Review by hanwavefront — CHANGES_REQUESTED** [review]
a couple minor things, otherwise lgtm

File: src/main/java/com/wavefront/sdk/common/logging/MessageSuppressingLogger.java [hunk 5]
```diff
@@ -58,21 +58,23 @@ public void log(Level level, String message) {
...
+        if (cname.endsWith("Logger")) {
+          lookingForLogger = false;
         }
       } else {
-        logRecord.setSourceClassName(className);
-        logRecord.setSourceMethodName(methodName);
-        return;
+        if (!cname.endsWith("Logger") && !cname.startsWith("java.lang.reflect.") &&
+            !cname.startsWith("sun.reflect.")
+            && !cname.startsWith("com.google.common.cache.LocalCache")) {
```
Comment (hanwavefront) [inline src/main/java/com/wavefront/sdk/common/logging/MessageSuppressingLogger.java:72]: nit: move && to end of previous line

**Review by hanwavefront — APPROVED** [review]
LGTM, just one nit

**Review by sushantdewan123 — APPROVED** [review]

**Pull request merged**